[GHF] Specify multiple mandatory checks

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73976
* #73975

Replace `mandatory_app_id` with `mandatory_checks_name`
Make `Facebook CLA Check` mandatory for all merge rules

malfet committed:
[GHF] Specify multiple mandatory checks

Replace `mandatory_app_id` with `mandatory_checks_name`
Make `Facebook CLA Check` mandatory for all merge rules

[ghstack-poisoned]

bigfootjon commented on .github/scripts/trymerge.py:
```diff
@@ -47,13 +47,33 @@
             }
             oid
-            checkSuites(filterBy: {appId: 12274}, first: 1) {
+            checkSuites(first: 50) {
```
The max page size in GitHub GraphQL is 100. I assume the bot has enough API tokens to make that work, and this would make this a little more future-proof

bigfootjon reviewed (approved):
LGTM besides the little nit about using `100` instead of `50`, also lint is failing

suo commented:
Can we do this through branch protections and required statuses? Is there an API to query those so that mergebot can respect them. That way we also get some UI from GH on required checks.

seemethere commented on .github/merge_rules.json:
```diff
@@ -15,42 +15,42 @@
       ],
       "approved_by": ["BowenBao", "garymm"],
-      "mandatory_app_id": 12274
+      "mandatory_checks_name": ["Facebook CLA Check"]
```
Maybe at this point it'd be better to have a script to generate this file since I feel like editing json directly is going to begin to get a bit cumbersome here?

seemethere reviewed (approved):
LGTM, just have a concern about editing json in the long run

janeyx99 commented:
> Can we do this through branch protections and required statuses? Is there an API to query those so that mergebot can respect them. That way we also get some UI from GH on required checks.

+1 to this, there is an API to get the required checks in both GraphQL and the rest api i believe

malfet committed:
Update on "[GHF] Specify multiple mandatory checks"


Replace `mandatory_app_id` with `mandatory_checks_name`
Make `Facebook CLA Check` mandatory for all merge rules

[ghstack-poisoned]

bigfootjon commented:
> Can we do this through branch protections and required statuses? Is there an API to query those so that mergebot can respect them. That way we also get some UI from GH on required checks.

Unfortunately we cannot. Mergebot pushes commits directly to `master` that do not have passing signals, so it would fail those requirements and the push would be rejected. @malfet attempted to do that in the initial design IIRC

malfet commented:
> Can we do this through branch protections and required statuses? Is there an API to query those so that mergebot can respect them. That way we also get some UI from GH on required checks.

There are two problems with this approach:
 - Mandatory Checks makes it impossible to push anything to the repo if one does not have administrative rights (which means, in order to merge ghstacked PR one would have to create service PRs against master, wait for those mandatory checks and then push)
 - Mandatory checks apply to entire repo, whereas with mergebot we want the flexibility to specify different set of checks for different rules (say docs rule makes only doc job mandatory, `ONNX` onnx_tests, etc)

malfet commented:
> +1 to this, there is an API to get the required checks in both GraphQL and the rest api. The REST API is https://docs.github.com/en/rest/reference/branches#get-status-checks-protection. The GraphQL is requiredStatusCheckContexts attribute of https://docs.github.com/en/graphql/reference/objects#branchprotectionrule

Yes, there is a way to query mandatory checks programmatically, but those are uniform across the repo

janeyx99 commented:
> * Mandatory Checks makes it impossible to push anything to the repo if one does not have administrative rights (which means, in order to merge ghstacked PR one would have to create service PRs against master, wait for those mandatory checks and then push)

Does this affect anything other than ghstack PRs? I'm also unsure about the administrative rights part--I thought the difference of ghstack PRs was their base branches/how does that connect with the admin rights?

> * Mandatory checks apply to entire repo, whereas with mergebot we want the flexibility to specify different set of checks for different rules (say docs rule makes only doc job mandatory, `ONNX` onnx_tests, etc)

In the longterm, I was thinking we would run all trunk tests before merging a PR to ensure greenness. This would be the simplest/most reliable, but it seems like you are adding some rudimentary target determination into mergebot then?

malfet commented:
@pytorchbot merge this

malfet commented:
> Does this affect anything other than ghstack PRs? I'm also unsure about the administrative rights part--I thought the difference of ghstack PRs was their base branches/how does that connect with the admin rights?
`git push` to a branch with protection rules can only be done by someone with admin rights. Regular users with write access to the repo can only "merged" PRs into protected branches. 

One potential workaround is to "commit" change to temporary branch, wait for mandatory checks to finish and then advance head to that branch

> In the longterm, I was thinking we would run all trunk tests before merging a PR to ensure greenness. This would be the simplest/most reliable, but it seems like you are adding some rudimentary target determination into mergebot then?
I thought in the long term we are going to have merge queue, and in order for PR to be qualified for mergequeue it should still go thru some subset of smoke tests, isn't it?
